receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go

```diff
@@ -16,54 +16,270 @@ package k8sapiserver
+				ci.ClusterNameKey: clusterName,
+				ci.MetricType:     ci.TypeClusterService,
+				ci.Timestamp:      timestampNs,
+				ci.TypeService:    service.ServiceName,
+				ci.K8sNamespace:   service.Namespace,
+				ci.Version:        "0",
+			}
+			if k.nodeName != "" {
+				attributes["NodeName"] = k.nodeName
+			}
+			md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+			result = append(result, md)
+		}
+
+		for namespace, podNum := range client.Pod.NamespaceToRunningPodNum() {
+			fields := map[string]interface{}{
+				"namespace_number_of_running_pods": podNum,
+			}
+			attributes := map[string]string{
+				ci.ClusterNameKey: clusterName,
+				ci.MetricType:     ci.TypeClusterNamespace,
+				ci.Timestamp:      timestampNs,
+				ci.K8sNamespace:   namespace,
+				ci.Version:        "0",
+			}
+			if k.nodeName != "" {
+				attributes["NodeName"] = k.nodeName
+			}
+			md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+			result = append(result, md)
+		}
+	}
 	return result
 }
 
-func (k *K8sAPIServer) start() error {
-	//TODO: add implementation
+func (k *K8sAPIServer) init() error {
+	var ctx context.Context
+	ctx, k.cancel = context.WithCancel(context.Background())
+
+	k.nodeName = os.Getenv("HOST_NAME")
+	if k.nodeName == "" {
+		return errors.New("missing environment variable HOST_NAME. Please check your deployment YAML config")
+	}
+
+	lockNamespace := os.Getenv("K8S_NAMESPACE")
+	if lockNamespace == "" {
+		return errors.New("missing environment variable K8S_NAMESPACE. Please check your deployment YAML config")
+	}
+
+	configMapInterface := k.k8sClient.ClientSet.CoreV1().ConfigMaps(lockNamespace)
+	if configMap, err := configMapInterface.Get(ctx, lockName, metav1.GetOptions{}); configMap == nil || err != nil {
+		k.logger.Info(fmt.Sprintf("Cannot get the leader config map: %v, try to create the config map...", err))
+		configMap, err = configMapInterface.Create(ctx,
+			&v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: lockNamespace,
+					Name:      lockName,
+				},
+			}, metav1.CreateOptions{})
+		k.logger.Info(fmt.Sprintf("configMap: %v, err: %v", configMap, err))
+	}
+
+	lock, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		lockNamespace, lockName,
+		k.k8sClient.ClientSet.CoreV1(),
+		k.k8sClient.ClientSet.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      k.nodeName,
+			EventRecorder: k.createRecorder(lockName, lockNamespace),
+		})
+	if err != nil {
+		k.logger.Warn("Failed to create resource lock", zap.Error(err))
+		return err
+	}
+
+	go k.startLeaderElection(ctx, lock)
+
 	return nil
 }
 
-// Stop stops the k8sApiServer
-func (k *K8sAPIServer) Stop() {
+// Shutdown stops the k8sApiServer
+func (k *K8sAPIServer) Shutdown() {
 	if k.cancel != nil {
 		k.cancel()
 	}
 }
+
+func (k *K8sAPIServer) startLeaderElection(ctx context.Context, lock resourcelock.Interface) {
+
+	for {
+		leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
```

Comment:
We're only computing metrics I think, why do we have leader election?

Reply:
More specifically - I'd expect metric dedupe to happen in the backend anyways, isn't duplicate metrics a standard thing when monitoring k8s?

Reply:
As collectors are deployed as daemon set for container insight, each collector will generate metrics independently of each other. Therefore we need to have leader election to make sure only one collector in a whole cluster is collecting cluster-level metrics. In theory, the backend can dedupe metrics (though I haven't heard that cloudwatch log backend can do that). However the cluster-level metrics are generated with different timestamps, even if they contains the same metrics values, they are actually different metrics (or emf logs) from the perspective of backend.

Reply:
Dedupe was probably the wrong word - these are just static metrics right? It's like an upload of k8s metrics to cloudwatch. Having 2 collectors uploading cluster metrics seems like a good idea (how long to recover from leader failure?) - but indeed maybe not 100.

It just feels like the design is too brute force. If we provide a yaml anyways with a dedicated lock namespace, Why not just deploy 1-2 collector instances into that namespace instead of coupling to the daemonset and dealing with the complexity of leader election?

Reply:
The leader election is provided by kubernetes api and should be mature enough for production usage. 

Having a dedicated collector for collecting only cluster-level metrics is definitely a design option. But this requires users to be aware of two deployments and change their perspective of how agent of container insights works. 

Considering that cloudwatch agent is using the leader election for collecting cluster-level metrics for container insights and there is no customer complaints about that, I think we should keep the current design.  To achieve parity between cloudwatch agent and otel collector for container insights, we also want to keep the deployment mode to be the same for smooth transition.